### PR TITLE
Updates extension to work with solidus

### PR DESCRIPTION
**Pull request by kenton**
Kept this as a few individual commits, so see those commit message for details, in particular, the ones that update `Spree::AmazonController` to work with some specifics of solidus.  I'd appreciate some extra eyes on that stuff.

Oh, and the extension this was forked from didn't have any specs.  Not sure if we want to add those now or add that as an ongoing dev task.  Thoughts?

### CODE REVIEW

File: app/controllers/spree/amazon_controller.rb [hunk 1]
```diff
@@ -111,8 +119,12 @@ def complete
...
     redirect_to root_path if @order.nil?
+
     while(@order.next) do
+    end
 
+    if @order.state == 'confirm'
+      @order.complete
     end
 
     if @order.completed?
```
Comment (mtomov) [inline app/controllers/spree/amazon_controller.rb:101]: Maybe if you can add a test or two for this controller.

File: README.md [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
 Installation
 ------------
 
-Add spree_amazon_payments to your Gemfile:
+Add solidus_amazon_payments to your Gemfile:
 
 ```
-gem 'spree_social', github: 'spree-contrib/spree_social', branch: '3-0-stable'
-gem 'spree_amazon_payments', github: 'amzn/spree_pay_with_amazon'
+gem 'solidus_social'
```
Comment (mtomov) [inline README.md:13]: Do we need social?

File: app/controllers/spree/amazon_controller.rb [hunk 3]
```diff
@@ -21,125 +20,127 @@ def address
...
+    payment.payment_method = Spree::Gateway::Amazon.first
+    payment.source ||= Spree::AmazonTransaction.create(
+      order_reference: params[:order_reference],
+      order_id: current_order.id
+    )
 
     payment.save!
 
-    render json: {}.to_json
+    render json: {}
```
Comment (mtomov) [inline app/controllers/spree/amazon_controller.rb:33]: haha, wow

File: app/models/spree_amazon/address.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                :address1, :phone
+
+  def initialize(attributes)
+    attributes.each_pair do |key, value|
+      send("#{key}=", value)
+    end
+  end
+
+  def first_name
+    unless name.blank?
```
Comment (mtomov) [inline app/models/spree_amazon/address.rb]: One can also terminate early with `return if name.blank?`

Comment (manmartinez) [inline app/models/spree_amazon/address.rb]: Ah good point, saves one line of code!

File: app/views/spree/amazon/confirm.html.erb [hunk 5]
```diff
@@ -12,7 +12,7 @@
   <%= render :partial => 'spree/amazon/order_details', :locals => { :order => @order } %>
 </fieldset>
 
-<%= form_tag amazon_order_complete_path(@order), :class => 'edit_order' do %>
+<%= form_tag complete_amazon_order_path, :class => 'edit_order' do %>
```
Comment (mtomov) [inline app/views/spree/amazon/confirm.html.erb]: do we maybe need to prefix this with `spree.complete_amazon_order_path`

File: app/controllers/spree/amazon_controller.rb [hunk 6]
```diff
@@ -21,125 +20,127 @@ def address
...
-    @order = Spree::Order.find_by(:number => params[:amazon_order_id])
+    @order = current_order
     authorize!(:edit, @order, cookies.signed[:guest_token])
 
-    redirect_to root_path if @order.nil?
     while(@order.next) do
+    end
 
+    if @order.state == 'confirm'
+      @order.complete
```
Comment (mtomov) [inline app/controllers/spree/amazon_controller.rb:98]: :+1:  good find

File: app/models/spree_amazon/address.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        return nil
+      end
+      new attributes_from_response(response.destination["PhysicalDestination"])
+    end
+
+    private
+
+    def mws(order_reference)
+      AmazonMws.new(order_reference, test_mode)
+    end
```
Comment (mtomov) [inline app/models/spree_amazon/address.rb:19]: Don't really understand the point of bringing this as a private method. It only adds a level of indirection from the main method.

If `preferred_test_mode` returns true/false, then naming the method with a question mark will be better - `in_test_mode?`.

**Comment by mtomov** [comment]
Yep, good stuff. Obviously you've spend quite some time on this : )

File: app/models/spree_amazon/address.rb (hunk 7 above)
Comment (manmartinez) [inline app/models/spree_amazon/address.rb:19]: The idea behind the `mws` method is to isolate the dependency on `AmazonMws`.

I'll rename `test_mode` good idea!

File: app/views/spree/amazon/confirm.html.erb (hunk 5 above)
Comment (manmartinez) [inline app/views/spree/amazon/confirm.html.erb]: Yeah I guess one can never be too safe

**Pull request merged**